### PR TITLE
Refactor: add ppadb instead of pyautogui

### PR DESCRIPTION
**Pull request by Soif2Sang**
This pull request refactors the bot's input and screenshot mechanisms to eliminate the PyAutoGUI dependency.

Key Changes:

- Input & Screenshots: The bot now uses the Android Debug Bridge (ADB) to send inputs and capture screenshots from the emulator. This removes all constraints related to screen size and window positioning.

- Headless Operation: The training process can now run headlessly, freeing up the user's mouse and monitor. This is a significant step towards scalability, allowing us to run multiple emulators simultaneously to accelerate the bot's learning.

- Image Updates: The previous training images were captured via Windows screenshots, leading to inconsistencies. This PR updates all training images to be captured directly from a Bluestacks instance, ensuring data quality.

### CODE REVIEW

**Comment by Soif2Sang** [comment]
> Have you been able to run multiple instances simultaneously?

No, it requires more modification to the repository.

I've looked a bit into it yesterday night and pytorch has a multiprocessing package which could suit for our needs, but I think we should split the PRs especially with such radical changes,

First release this PR if you think it is mature enough and after that release a PR to scale the training process

**Comment by Shaarav4795** [comment]
Can confirm this works... Tested myself. Just wanted to say that this project works perfectly on macOS as well, as I have tested it on there. I wasn't able to download the pyadb-reborn library, so I went for the original, less frequently updated one.

**Comment by Soif2Sang** [comment]
> > Have you been able to run multiple instances simultaneously?
> 
> No, it requires more modification to the repository.
> 
> I've looked a bit into it yesterday night and pytorch has a multiprocessing package which could suit for our needs, but I think we should split the PRs especially with such radical changes,
> 
> First release this PR if you think it is mature enough and after that release a PR to scale the training process

I'm currently testing pytorch multiprocessing, It looks to work perfectly fine with fine tunning

**Comment by Soif2Sang** [comment]
> Can confirm this works... Tested myself. Just wanted to say that this project works perfectly on macOS as well, as I have tested it on there. I wasn't able to download the pyadb-reborn library, so I went for the original, less frequently updated one.

I haven't tested on MacOS, try to add adb in your env path, then try `setup_adb.py`

**Comment by Shaarav4795** [comment]
> > Can confirm this works... Tested myself. Just wanted to say that this project works perfectly on macOS as well, as I have tested it on there. I wasn't able to download the pyadb-reborn library, so I went for the original, less frequently updated one.
> 
> I haven't tested on MacOS, try to add adb in your env path, then try `setup_adb.py`

I had to install the adb packages with home brew, and after that it worked. Is it intentional that the the terminal says pressing okay at the end of the game, but never actually does?

**Comment by Soif2Sang** [comment]
> > > Can confirm this works... Tested myself. Just wanted to say that this project works perfectly on macOS as well, as I have tested it on there. I wasn't able to download the pyadb-reborn library, so I went for the original, less frequently updated one.
> > 
> > 
> > I haven't tested on MacOS, try to add adb in your env path, then try `setup_adb.py`
> 
> I had to install the adb packages with home brew, and after that it worked. Is it intentional that the the terminal says pressing okay at the end of the game, but never actually does?

Definitely not, it should tap on the "ok" button

**Comment by craftguy1243** [comment]
Before multi instance training is added we should update the training logic. Right now the bot is learning to play the game the wrong way. I don't know how easy it would be to add card type detection and counters, as well as cycle counting, but if that was implemented alongside multiple people running multi instance training the bot would get pretty decent at cr much faster. Also later on if possible elixr management would be huge.

**Comment by Soif2Sang** [comment]
> wrong way. I don't know how easy it would be to add card type detection and counters, as well as cycle counting, but if that was implemented alongside multiple people running multi instance training the bot would get pretty decent at cr much faster. Also later on if possible elixr management would be huge.

I would love to help the implémentation of it, I'm not new to python but I'm really *new* to AI, I've thought about changing the state, implementing HP tower vision. I'm definitely capable of doing that

**Comment by Soif2Sang** [comment]
> > > wrong way. I don't know how easy it would be to add card type detection and counters, as well as cycle counting, but if that was implemented alongside multiple people running multi instance training the bot would get pretty decent at cr much faster. Also later on if possible elixr management would be huge.
> > 
> > 
> > I would love to help the implémentation of it, I'm not new to python but I'm really _new_ to AI, I've thought about changing the state, implementing HP tower vision. I'm definitely capable of doing that
> 
> I've never even heard of roboflow before so I'm not too sure what AI stuff is going on with it either. I wonder if it would be possible to add card recognition and then feed human played games into the epsilon instead of adding hardcoded counters. If you just feed the bot all of the correct actions with positive reinforcement for a scenario it would improve learning speed and skill immensely. Kind of like a neural net. That would be pretty complicated though and probably not entirely feasible. If it did work perfectly you could have a bot with top 1% skill 😀

The hardest thing is to create a robust dataset that allows the AI to label entities on the screen with high accuracy, I took a look at roboflow available dataset, and I've fond some with around 3k unlabeled images, labeling image is a nightmare to do from what i've seen

**Comment by craftguy1243** [comment]
> > > > wrong way. I don't know how easy it would be to add card type detection and counters, as well as cycle counting, but if that was implemented alongside multiple people running multi instance training the bot would get pretty decent at cr much faster. Also later on if possible elixr management would be huge.
> > > 
> > > 
> > > I would love to help the implémentation of it, I'm not new to python but I'm really _new_ to AI, I've thought about changing the state, implementing HP tower vision. I'm definitely capable of doing that
> > 
> > 
> > I've never even heard of roboflow before so I'm not too sure what AI stuff is going on with it either. I wonder if it would be possible to add card recognition and then feed human played games into the epsilon instead of adding hardcoded counters. If you just feed the bot all of the correct actions with positive reinforcement for a scenario it would improve learning speed and skill immensely. Kind of like a neural net. That would be pretty complicated though and probably not entirely feasible. If it did work perfectly you could have a bot with top 1% skill 😀
> 
> The hardest thing is to create a robust dataset that allows the AI to label entities on the screen with high accuracy, I took a look at roboflow available dataset, and I've fond some with around 3k unlabeled images, labeling image is a nightmare to do from what i've seen

Do you just have to manually label each image? 2 people could probably do it in a month, or something with chatgpt api that auto labels. If labeling meant fully working bot learning from human gameplay I would do it.

**Comment by Shaarav4795** [comment]
Agree with this. Would definitely put my time into this if it meant a very capable bot 😅On 9 Aug 2025, at 1:29 pm, craftguy1243 ***@***.***> wrote:﻿craftguy1243 left a comment (krazyness/CRBot-public#8)




wrong way. I don't know how easy it would be to add card type detection and counters, as well as cycle counting, but if that was implemented alongside multiple people running multi instance training the bot would get pretty decent at cr much faster. Also later on if possible elixr management would be huge.

I would love to help the implémentation of it, I'm not new to python but I'm really new to AI, I've thought about changing the state, implementing HP tower vision. I'm definitely capable of doing that

I've never even heard of roboflow before so I'm not too sure what AI stuff is going on with it either. I wonder if it would be possible to add card recognition and then feed human played games into the epsilon instead of adding hardcoded counters. If you just feed the bot all of the correct actions with positive reinforcement for a scenario it would improve learning speed and skill immensely. Kind of like a neural net. That would be pretty complicated though and probably not entirely feasible. If it did work perfectly you could have a bot with top 1% skill 😀

The hardest thing is to create a robust dataset that allows the AI to label entities on the screen with high accuracy, I took a look at roboflow available dataset, and I've fond some with around 3k unlabeled images, labeling image is a nightmare to do from what i've seen

Do you just have to manually label each image? 2 people could probably do it in a month, or something with chatgpt api that auto labels. If labeling meant fully working bot learning from human gameplay I would do it.

**Comment by Soif2Sang** [comment]
I'm very new to AI and I'm unsure if 3k images will be enough to make a robust enough recognition model
In case it's enough we will still need a ton of training, maybe thousands of games to produce a good result ar above few thousands trophées
I think you should not expect top 1% performance 
I've forked a roboflow workflow and started labeling images, it takes time.. I'll try my best to do it during my free time, I've got like two weeks of holidays, but there's also the battlefield 6 beta 😂

**Comment by craftguy1243** [comment]
> I'm very new to AI and I'm unsure if 3k images will be enough to make a robust enough recognition model
> 
> In case it's enough we will still need a ton of training, maybe thousands of games to produce a good result ar above few thousands trophées
> 
> I think you should not expect top 1% performance 
> 
> I've forked a roboflow workflow and started labeling images, it takes time.. I'll try my best to do it during my free time, I've got like two weeks of holidays, but there's also the battlefield 6 beta 😂
> 
> 

Share the link I'll help. I think the best path to take would be watching real people play as it would be fairly easy for the bot to just constantly watch top games from tv royale or you playing. There really isn't a ruleset you can use to train a cr bot that will be good because of the millions of edge cases.
